ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision/BulkSync.hs

```diff
@@ -0,0 +1,613 @@
+                -- We chose the same peer as before but there were no more
+                -- blocks in flight for this peer: record a new start time.
+                peersOrderStart <- getMonotonicTime
+                writePeersOrder $ PeersOrder {peersOrderCurrent, peersOrderStart, peersOrderOthers}
+          _ -> pure ()
+
+      peerHasBlocksInFlight peer =
+        case find (\(_, (_, _, _, peer', _)) -> peer == peer') candidatesAndPeers of
+          Just (_, (_, inflight, _, _, _)) -> not $ Map.null $ peerFetchBlocksInFlight inflight
+          Nothing -> error "blocksInFlightForPeer"
+
+snoc :: [a] -> a -> [a]
+snoc [] a = [a]
+snoc (x : xs) a = x : snoc xs a
+
+-- | Given a list of candidate fragments and their associated peers, choose what
+-- to sync from who in the bulk sync mode.
+fetchDecisionsBulkSync ::
+  ( HasHeader header,
+    HeaderHash header ~ HeaderHash block,
+    Eq peer
+  ) =>
+  FetchDecisionPolicy header ->
+  -- | The current chain, anchored at the immutable tip.
+  AnchoredFragment header ->
+  (Point block -> Bool) ->
+  MaxSlotNo ->
+  PeersOrder peer ->
+  -- | Association list of the candidate fragments and their associated peers.
+  -- The candidate fragments are anchored in the current chain (not necessarily
+  -- at the tip; and not necessarily forking off immediately).
+  [(AnchoredFragment header, PeerInfo header peer extra)] ->
+  -- | Association list of the requests and their associated peers. There is at
+  -- most one accepted request; everything else is declined. Morally, this is a
+  -- map from peers to @'FetchDecision' ('FetchRequest' header)@ with at most
+  -- one @'FetchRequest' header@.
+  ( Maybe (FetchRequest header, PeerInfo header peer extra),
+    [(FetchDecline, PeerInfo header peer extra)]
+  )
+fetchDecisionsBulkSync
+  fetchDecisionPolicy
+  currentChain
+  fetchedBlocks
+  fetchedMaxSlotNo
+  peersOrder
+  candidatesAndPeers = combineWithDeclined $ do
+    -- Step 1: Select the candidate to sync from. This already eliminates peers
+    -- that have an implausible candidate. It returns the remaining candidates
+    -- (with their corresponding peer) as suffixes of the immutable tip.
+    ( theCandidate :: ChainSuffix header,
+      candidatesAndPeers' :: [(ChainSuffix header, PeerInfo header peer extra)]
+      ) <-
+      MaybeT $
+        selectTheCandidate
+          fetchDecisionPolicy
+          currentChain
+          candidatesAndPeers
+
+    -- Step 2: Filter out from the chosen candidate fragment the blocks that
+    -- have already been downloaded, or that have a request in flight (except
+    -- for the requests in flight that are ignored). NOTE: if not declined,
+    -- @theFragments@ is guaranteed to be non-empty.
+    let (theFragments :: FetchDecision (CandidateFragments header)) =
+          pure theCandidate
+            >>= filterNotAlreadyFetched fetchedBlocks fetchedMaxSlotNo
+            >>= filterNotAlreadyInFlightWithAnyPeerNonIgnored candidatesAndPeers
```

Comment:
It aches me to read these names. I'd find more natural to write:
```haskell
  >>= dropAlreadyFetched fetchedBlocks fetchedMaxSlotNo
￼ >>= dropAlreadyInFlightWithAnyNonIgnoredPeer candidatesAndPeers
```

Reply:
I agree very much. The second one has been removed entirely after our discussion in the sync. The first one pre-dates our effort and that is why I haven't renamed it. I can, however, while we're at it.